sys/include/wtimer.h

```diff
@@ -0,0 +1,330 @@
+ * @return 0 on success, < 0 on error
+ */
+int wtimer_usleep_until(uint32_t *last_wakeup, uint32_t usecs);
+
+/**
+ * @brief get the current system time as 64bit microsecond value
+ *
+ * @return current time as 64bit microsecond value
+ */
+uint64_t wtimer_now64(void);
+
+/**
+ * @brief get the current system time into a timex_t
+ *
+ * @param[out] out pointer to timex_t the time will be written to
+ */
+void wtimer_now_timex(timex_t *out);
+
+/**
+ * @brief Set a timer that sends a message
+ *
+ * This function sets a timer that will send a message <offset> microseconds from now.
+ *
+ * The mesage struct specified by msg parameter will not be copied, e.g., it
+ * needs to point to valid memory until the message has been delivered.
+ *
+ *
+ * @param[in] timer         timer struct to work with
+ * @param[in] offset        microseconds from now
+ * @param[in] msg           ptr to msg that will be sent
+ * @param[in] target_pid    pid the message will be sent to
+ */
+void wtimer_set_msg(wtimer_t *timer, uint32_t offset, msg_t *msg, kernel_pid_t target_pid);
+
+/**
+ * @brief Set a timer to execute a callback at some time in the future
+ *
+ * Expects timer->callback to be set.
+ *
+ * The callback specified in the timer struct will be executed <offset> microseconds in
+ * the future.
+ *
+ * BEWARE! Callbacks from wtimer_set() are being executed in interrupt context
+ * (unless offset < WTIMER_BACKOFF). DON'T USE THIS FUNCTION unless you know
+ * *exactly* what that means.
+ *
+ * @param[in] timer the timer structure to use
+ * @param[in] offset time in microseconds from now specifying that timer's callback's execution time
+ */
+void wtimer_set(wtimer_t *timer, uint32_t offset);
+
+/**
+ * @brief wtimer backoff value
+ *
+ * All timers that are less than WTIMER_BACKOFF microseconds in the future will
+ * just spin.
+ *
+ * This is supposed to be defined per-device in e.g., periph_conf.h.
+ */
+#ifndef WTIMER_BACKOFF
+#define WTIMER_BACKOFF 30
+#endif
+
+/**
+ * @brief wtimer overhead value
+ *
+ * This value specifies the time a timer will be late if uncorrected.
+ *
+ * E.g., with WTIMER_OVERHEAD == 0
+ * now=wtimer_now();
+ * wtimer_set(&timer, X);
+ * overhead=X-now
+ *
+ * wtimer automatically substracts WTIMER_OVERHEAD from a timer's target time.
+ *
+ * This is supposed to be defined per-device in e.g., periph_conf.h.
+ */
+#ifndef WTIMER_OVERHEAD
+#define WTIMER_OVERHEAD 20
+#endif
+
+#ifndef WTIMER_ISR_BACKOFF
+#define WTIMER_ISR_BACKOFF 20
+#endif
+
+#ifndef WTIMER_MASK
+/**
+ * @brief wtimer timer mask
+ *
+ * This value specifies the mask relative to 0xffffffff that the used timer
+ * counts to, e.g., 0xffffffff & ~TIMER_MAXVALUE.
+ *
+ * For a 16bit timer, the mask would be 0xFFFF0000, for a 24bit timer, the mask
+ * would be 0xFF000000. Don't set this for 32bit timers.
+ *
+ * This is supposed to be defined per-device in e.g., periph_conf.h.
+ */
+#define WTIMER_MASK 0
```

Comment:
The logic here is kind of reversed compared to most other bitmasks I have seen around the tree where the used bits are set to 1 and the unused bits are zeroed. (which lets you do `x & MYMASK`, instead of `x & ~(MYMASK)` to get the value you need)